### ISSUE TITLE
Proposal for conditionals

### ISSUE THREAD

**Issue by tetron** [common-workflow-language/common-workflow-language#494]
Here's my most recent design sketch for conditionals in CWL.  Next step is will be to implement the necessary schema changes in a PR to the v1.1-dev1 spec and do a trial implementation in cwltool (with --enable-dev flag).

The general form is a "switch" statement with one or more conditional cases and a default case.

A "case" is an input parameter reference or an expression which is coerced to a boolean based on Javascript rules.  If there is a list of cases, they are evaluated in the order and the first "true" case is followed.  A "default" case is required, and is followed if all conditional cases are false.

In addition, this introduces is a new "result" field on a step.  In its general form, this specifies how to construct the output object for the step.  If no "run" is specified, the "result" can be used to construct a object from constant values or propagate values from input to output.  (The combination of "valueFrom" on steps inputs and "result" for step outputs may supercede ExpressionTool for many uses).

condensed form:
```
  step2:
    in:
      a: a
    out: [out]
    switch:
      "$(inputs.a > 2)": blah.cwl
      default:
        result:
          out: 12
```

expanded form:
```
  step1:
    in:
      a: a
    out: [out]
    switch:
      - case: $(inputs.a > 2)
        run: blah.cwl
      - case: default
        result:
          - resultParam: out
            valueFrom: 12
```

Comments, feedback appreciated.

Thanks,
Peter

**Comment by stain** [comment]
It's beautiful! I like that `default` is required.

**Comment by stain** [comment]
What is the ordering in the condensed form? Yaml dictionaries are not ordered. Does condensed form only permit a single conditional + `default`?

**Comment by tetron** [comment]
A slight tweak to the behavior I've been considering is that if `default` isn't provided and none of the conditionals pass, that is a workflow step failure.  What I want to avoid is a default `default` behavior that might put you on an unanticipated path through the graph.

I don't know if we want a `ConditionalRequirement`, the feeling has been that the boilerplate feature flags in v1.0 don't add much value, it should be the responsibility of engines to detect and report features they don't support.

The generic expansion of maps to lists does lexical sorting by key, although I don't think that behavior is explicitly specified in schema salad.

**Comment by jmchilton** [comment]
I'm not +1 or -1 on this. I agree it is very clean and it looks like it simplifies some questions about walking down alternative sets of steps for implementers - you are always walking down all the steps and that is kind of nice - if I conditionally invokes some steps they need to be a subworkflow. This is potentially the cleanest a useful syntax could be in terms of implementing it in Galaxy. I'm not sure that is a great thing to optimize though.

For some context about this - I'm trying to develop an opinion about workflow scattering. In particular I don't think it is possible to twice scatter a workflow in CWL directly right? If I have a tool and I want to run a array of a array of files over it - I need to create a subworkflow for that tool that scatters and then scatter over the subworkflow itself. (Correct me if I am wrong.) My intuition says it is bad that you need subworkflows for this feature and that subworkflow expansion isn't just like sugar - you gain the ability to do new things with subworkflows that aren't possible without them. So I'm starting to think of "forces subworkflows" as a sort of language smell for CWL. That said - it is all very clean and you don't have to make some assumptions that I made in Galaxy to implement double nesting and such - the users intent is less ambiguous in some corner cases because scattering is always one level deep. 

This particular syntax has that same smell I think - there are things you cannot do without subworfklows with this syntax. People will be forced to create subworkflows for things they might not want to be subworkflows.

Regardless - I guess the proposal should be updated to discuss the semantics of scattering a step with a switch in it. I can't think of any complications off the top of my head - but probably best to be explicit about whether different scattered inputs can walk down different cases or if the case logic applies to the pre-scattered input (I assume it is the former right?).

**Comment by tetron** [comment]
@LourensVeen Thanks for the feedback!  I had a similar thought and going back in the mailing list there should be a post with a similar proposal (I'll have to go dig it up).  I feel like the community reaction was negative on that approach, but since nothing is finalized (or implemented) it would be worth reconsidering.  I agree it is more elegant in some ways.

**Comment by tetron** [comment]
Here it is https://groups.google.com/forum/m/?source=mog&gl=us#!topic/common-workflow-language/JU7PSEKr97M

**Comment by LourensVeen** [comment]
Ah, thanks! I only recently found out about the Google Group, and I haven't read all the way back in the archives. Yes, you're right, what I'm proposing above is exactly equivalent to your original suggestion in that thread. That proves it's a good idea, right? :)

Quite a few people in that thread seem to be expecting a switch statement, but I'm not really seeing anyone mention an issue that as actually an issue with your original proposal. Karl suggests fixing the multiple inputs/outputs problem by listing them all, but that actually gets a lot more complex syntactically than having separate steps. I didn't read everything completely, but I don't think anyone else addresses it.

It's possible of course to move the conditionals to the output step, adding an expression there, if people want something that looks like a switch statement (just to be clear, I'm assuming no fall-through):

```
steps:
  step1a:
    in:
      a: a
    out: [out]
    run: blah.cwl

  step1b:
    in:
      b: a
    out: [out2]
    run: blah2.cwl

outputs:
  out:
    valueFrom: $(
      case inputs.a > 2: step1a.out
      case inputs.a == 1: step1b.out
      default: 12
    )
```

Or pick whatever syntax you like. Perhaps even

```
outputs:
  out:
    switch:
      "inputs.a > 2": step1a.out
      "inputs.a == 1": step1b.out
      default: 12
```

The problem in my eyes with all the other proposals is that we end up adding all sorts of stuff on the _inside_ of a WorkflowStep to allow describing how multiple possible processes inside of it are related, while we already have a perfectly good system for describing relationships between processes on the _outside_ of WorkflowStep. So everything gets doubled up, and that's bad.

**Comment by tetron** [comment]
Another idea: make out-of-band signaling of step status explicit.  Currently has "success" and "fail" but could also have "conditional skip".

**Comment by stain** [comment]
conditional skip is more flexible - but  without the explicit `default` and the conditional being done on the "outside" rather within a single step (which still complies to the outer dataflow semantics), is more dangerous and easy to break the dataflow. 

That's how people would make their own conditionals in early Taverna 1, using ["Always fail" helpers](http://www.mygrid.org.uk/usermanual1.7/conditional_branching.html). It turned out people were not very good at writing the boolean tests. While they intended a traditional if-else they used:

* A: fail if `i > 2`
* B: fail if `i < 2`

And got as expected:

* `i=1`: A fails, B passes
* `i=3`: A passes, B fails

..they would often not cover all the possibilities in such split boolean tests:

* `i=2`: A passes, B passes  (`merge_first`: one of the values picked at random)
* `i=2`: A passes, B passes (`merge`: both values merged into list - which is first?)
* `i="fred"`: A fails, B fails  (`merge_first`: downstream won't get any outputs at all?)
* `i="fred"`: A fails, B fails  (`merge`: empty list)

and thus there would be 6 possible workflow states from what was meant as a simple `if-else`. This can be hard enough to consider for experienced programmers - particularly if checking two different parameters! I would still advocate for a explicit and required `default` and capture it within a step. It is true this might mean parameter mapping, but that is already supported by `run`.

In Taverna we added the possible return type of [Error documents](http://dev.mygrid.org.uk/wiki/display/developer/Differences+between+Taverna+1+and+Taverna+2+workflows#DifferencesbetweenTaverna1andTaverna2workflows-Errordocuments) as a replacement for values -- this is very powerful workflow semantics - e.g. a list can contain an error, or you can scatter-iterate over an error document with list depth (And get new lower depth error documents which keeps the "Caused by" trace) -- However I am not so sure about enforcing such semantics on all CWL implementations; although we currently are a bit too quiet on error handling.

In Taverna we experimented with the idea of propagating rich `null` values instead of error documents - kind of positive errors.  This is kind of what you are proposing with a `conditional skip` status - it's not failed exceptionally, but deliberately not run, a kind of positive error document.   I think you could do that in your original proposal by simply returning `null` -- we can have a `skip` or `null` syntactic sugar to avoid having to list that for each output.  (The output types would still need to be `null`-compatible downstream)

**Comment by tetron** [comment]
Here's a different direction.  CWL documents are inputs to the workflow step, and `run` may consist of an expression which selects what to run and returns either a single process or list of processes to execute, or an immediate output object:

```
steps:
  step2:
    in:
      a: a
      branch: {default: {class: File, location: branch.cwl}}
    out: [out]
    run: |
      ${
        if (inputs.a > 2) {
          return {job: inputs, tool: inputs.branch};
        } else {
          return {output: {result: 12}};
        }
      }
```

**Comment by LourensVeen** [comment]
I think @stain's point is very convincing, that it's better to keep the topology between the steps static, and do the switch on the inside. Also the case for a case rather than an if is also strong, many people seem to like that idea also in the previous thread linked above.

I'm not sure that I like using a piece of script to do it though. Maybe something like this?
```
steps:
  step2:
    in:
      a: a
      branch: {default: {class: File, location: branch.cwl}}
    out: [out]
    switch:
      - case: inputs.a > 2
        run: $(inputs.branch)
      - run:
          class: ExpressionTool
          outputs: { outputs: {out: 12} }
```
Not sure if that ExpressionTool is entirely valid :).

**Comment by kaushik-work** [comment]
@pvanheus could you expand on this a bit for a non PLT person like me :) Also would you support the idea of `switch` statements, rather than just a conditional? I was looking over some use cases from cancer genomics pipelines and from some other private client pipelines, and they have a strong need for a `switch` statement, where one of several sub-workflows are executed depending on an input value.

**Comment by stain** [comment]
Conditionals was implemented in CWL 1.2, see [`when` on step](https://www.commonwl.org/v1.2/Workflow.html#Conditional_execution_(Optional)) along with [`pickValue` on a step input](https://www.commonwl.org/v1.2/Workflow.html#Picking_non-null_values_among_inbound_data_links)
